File: clients/kmp/shared/src/commonTest/kotlin/org/wyrdsekai/app/engine/InMemoryEventJournal.kt
package org.wyrdsekai.app.engine

import org.wyrdsekai.app.engine.event.WorldEvent
import org.wyrdsekai.app.engine.persistence.EventJournal

/** Test double for EventJournal — in-memory, no platform deps. */
class InMemoryEventJournal : EventJournal {
    private val events = mutableMapOf<String, MutableList<WorldEvent>>()
    private val snapshots = mutableMapOf<String, String>()
    private val appendCounts = mutableMapOf<String, Int>()

    /** Auto-compaction threshold. Set to 0 to disable. */
    var autoCompactThreshold: Int = 0

    override suspend fun append(roomId: String, event: WorldEvent) {
        events.getOrPut(roomId) { mutableListOf() }.add(event)

        if (autoCompactThreshold > 0) {
            val count = (appendCounts[roomId] ?: 0) + 1
            appendCounts[roomId] = count
            if (count >= autoCompactThreshold) {
                compact(roomId)
                appendCounts[roomId] = 0
            }
        }
    }

    override suspend fun replay(roomId: String): List<WorldEvent> =
        events[roomId]?.toList() ?: emptyList()

    override suspend fun saveSnapshot(roomId: String, snapshotJson: String) {
        snapshots[roomId] = snapshotJson
    }

    override suspend fun loadSnapshot(roomId: String): String? = snapshots[roomId]

    override suspend fun compact(roomId: String, keepLast: Int) {
        val list = events[roomId] ?: return
        if (list.size > keepLast) {
            val trimmed = list.takeLast(keepLast).toMutableList()
            events[roomId] = trimmed
        }
    }

    fun eventCount(roomId: String): Int = events[roomId]?.size ?: 0
    fun allEvents(roomId: String): List<WorldEvent> = events[roomId]?.toList() ?: emptyList()
    fun clear() { events.clear(); snapshots.clear(); appendCounts.clear() }
}

File: clients/rn/src/engine/persistence/InMemoryEventJournal.ts
/**
 * In-memory EventJournal — no platform deps, useful for tests and mobile default.
 * TypeScript port of KMP's InMemoryEventJournal.kt.
 */

import type { WorldEvent } from '../events/WorldEvent';
import type { EventJournal } from './EventJournal';

export class InMemoryEventJournal implements EventJournal {
  private events = new Map<string, WorldEvent[]>();
  private snapshots = new Map<string, string>();

  async append(roomId: string, event: WorldEvent): Promise<void> {
    const list = this.events.get(roomId) ?? [];
    list.push(event);
    this.events.set(roomId, list);
  }

  async replay(roomId: string): Promise<WorldEvent[]> {
    return [...(this.events.get(roomId) ?? [])];
  }

  async saveSnapshot(roomId: string, snapshotJson: string): Promise<void> {
    this.snapshots.set(roomId, snapshotJson);
  }

  async loadSnapshot(roomId: string): Promise<string | null> {
    return this.snapshots.get(roomId) ?? null;
  }

  async compact(roomId: string, keepLast = 500): Promise<void> {
    const list = this.events.get(roomId);
    if (!list || list.length <= keepLast) return;
    this.events.set(roomId, list.slice(-keepLast));
  }

  async getEventCount(roomId: string): Promise<number> {
    return this.events.get(roomId)?.length ?? 0;
  }

  /** Synchronous event count — convenience for tests. */
  eventCount(roomId: string): number {
    return this.events.get(roomId)?.length ?? 0;
  }

  allEvents(roomId: string): WorldEvent[] {
    return [...(this.events.get(roomId) ?? [])];
  }

  clear(): void {
    this.events.clear();
    this.snapshots.clear();
  }
}
